pkg/skaffold/cluster/minikube.go

```diff
@@ -0,0 +1,206 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cluster
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"log"
+	"net/url"
+	"os"
+	"os/exec"
+
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/rest"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/constants"
+	k8s "github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/client"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubernetes/context"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+)
+
+var GetClient = getClient
+
+// To override during tests
+var (
+	mkBinaryFunc  = minikubeBinary
+	k8sConfigFunc = context.GetRestClientConfig
+)
+
+type Client interface {
+	// IsMinikube returns true if the given kubeContext maps to a minikube cluster
+	IsMinikube(kubeContext string) bool
+	// MinikubeExec returns the Cmd struct to execute minikube with given arguments
+	MinikubeExec(arg ...string) (*exec.Cmd, error)
+}
+
+type clientImpl struct{}
+
+func getClient() Client {
+	return clientImpl{}
+}
+
+func (clientImpl) IsMinikube(kubeContext string) bool {
+	// short circuit if context is 'minikube'
+	if kubeContext == constants.DefaultMinikubeContext {
+		return true
+	}
+	_, err := mkBinaryFunc()
+	if err != nil {
+		return false // minikube binary not found
+	}
+
+	if ok, err := matchNodeLabel(kubeContext); err != nil {
+		logrus.Debugf("failed to check minikube node labels: %v", err)
+		return false
+	} else if ok {
+		logrus.Debugf("Detected minikube cluster for context %s due to matched labels", kubeContext)
+		return true
+	}
+
+	if ok, err := matchProfileAndServerURL(kubeContext); err != nil {
+		logrus.Debugf("failed to match minikube profile: %v", err)
+		return false
+	} else if ok {
+		logrus.Debugf("Detected minikube cluster for context %s due to matched profile name or server url", kubeContext)
+		return true
+	}
+	return false
+}
+
+func (clientImpl) MinikubeExec(arg ...string) (*exec.Cmd, error) {
+	return minikubeExec(arg...)
+}
+
+func minikubeExec(arg ...string) (*exec.Cmd, error) {
+	b, err := mkBinaryFunc()
+	if err != nil {
+		return nil, fmt.Errorf("getting minikube executable: %w", err)
+	}
+	return exec.Command(b, arg...), nil
+}
+
+func minikubeBinary() (string, error) {
+	execName := "minikube"
+	if found, _ := util.DetectWSL(); found {
+		execName = "minikube.exe"
+	}
+	filename, err := exec.LookPath(execName)
+	if err != nil {
+		return "", errors.New("unable to find minikube executable. Please add it to PATH environment variable")
+	}
+	if _, err := os.Stat(filename); os.IsNotExist(err) {
```

Comment:
I wonder when this would happen!

Reply:
this is actually [existing code](https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/docker/client.go#L181) that I just relocated. So I didn't try to delve too deeply into the implementation

Reply:
I guess it'd happen if you have the minikube binary in `$PWD` and not on `$PATH`.

Reply:
ya, I think that's the case
```
❯ cat main.go
package main

import (
        "fmt"
        "os"
)

func main() {
        if _, err := os.Stat("minikube"); os.IsNotExist(err) {
                fmt.Println("Not found!")
        } else {
                fmt.Println("Found!")
        }
}
❯ go run main.go
Not found!
❯ touch minikube
❯ go run main.go
Found!
```